### PR TITLE
jax.make_mesh: re-add multi-slice error on GPU

### PR DESCRIPTION
**Pull request by olupton**
Also update the multi-process docs and dull a sharp edge on GPU by sorting by device ID.

This reverts commit 2cc6522a5629ff872e7453656e4d7f8f363f9d84.

### CODE REVIEW

File: jax/_src/mesh_utils.py [hunk 1]
```diff
@@ -846,6 +846,13 @@ def create_device_mesh(
         allow_split_physical_axes=allow_split_physical_axes,
     )
     return device_mesh
+  elif last_device.platform == 'gpu':
```
Comment (yashk2810) [inline jax/_src/mesh_utils.py:849]: multi-slice device list can't be passed here since make_mesh errors out now right?

So no-need for this edge-case?

File: docs/multi_process.md [hunk 2]
```diff
@@ -421,7 +417,61 @@ because it's simpler but also because it can choose more performant device
...
+| DCN     | InfiniBand (IB), Ethernet, EFA, TCPXO, ... |
+| Slice   | NVLink domain (e.g. 18 hosts with 4 GPUs each in a GB200-NVL72 rack-scale system, 1 host with 8 GPUs in an HGX B200 NVL8 system, ...) |
+
+JAX will automatically detect which devices belong to which slices  during
+{func}`jax.distributed.initialize` and assign `slice_index` values to the
+devices accordingly.
+
+If you are using devices within a single slice (NVLink domain), it is sufficient to use
+{func}`jax.make_mesh` and, if you need a little more control,
+{func}`jax.experimental.mesh_utils.create_device_mesh`.
```
Comment (yashk2810) [inline docs/multi_process.md]: Remove the mention of `create_device_mesh` from here.

File: docs/multi_process.md [hunk 3]
```diff
@@ -421,7 +417,61 @@ because it's simpler but also because it can choose more performant device
...
+Mesh(create_hybrid_device_mesh((1, devices_per_slice), (num_slices, 1)), axis_names=("dcn", "ici"))
+```
+will produce a mesh that maps the `dcn` axis to DCN interconnect and the `ici` axis to
+ICI interconnect.
+```{warning}
+On popular GPU systems such as NVIDIA DGX H100 (A3 instances on GCP, P5 on AWS, ...),
+each node is a single NVLink domain, so even a two-node job is a multi-slice device mesh.
+```
+
+By default, `jax.devices()` is ordered by the process indices assigned when distributed
```
Comment (yashk2810) [inline docs/multi_process.md]: Please revert this entire paragraph

File: docs/multi_process.md [hunk 4]
```diff
@@ -421,7 +417,61 @@ because it's simpler but also because it can choose more performant device
...
+
+JAX will automatically detect which devices belong to which slices  during
+{func}`jax.distributed.initialize` and assign `slice_index` values to the
+devices accordingly.
+
+If you are using devices within a single slice (NVLink domain), it is sufficient to use
+{func}`jax.make_mesh` and, if you need a little more control,
+{func}`jax.experimental.mesh_utils.create_device_mesh`.
+
+If you are using a larger mesh of devices that spans multiple slices, JAX encourages
```
Comment (yashk2810) [inline docs/multi_process.md]: "If you are using devices that span multiple slices, use {func}`jax.experimental.mesh_utils.create_hybrid_device_mesh`."

No need to mention jax.make_mesh will raise an error since that's an implementation detail which will probably change in the future.

**Review by yashk2810 — CHANGES_REQUESTED** [review]

File: jax/_src/mesh_utils.py (hunk 1 above)
Comment (olupton) [reply jax/_src/mesh_utils.py:849]: You can still call `jax.experimental.mesh_utils.create_device_mesh` directly.

I would like to do `<something>` to remove the sharp edge that it [says](https://docs.jax.dev/en/latest/_autosummary/jax.experimental.mesh_utils.create_device_mesh.html) it produces a performant mesh, but if you run on multi-slice GPU then whether it does or not is dependent on how your job scheduler distributes jobs.

Here I tried to fix it; the alternatives I saw were an explicit error even if you call it directly, or more explicit documentation.

Comment (yashk2810) [reply jax/_src/mesh_utils.py:849]: > You can still call jax.experimental.mesh_utils.create_device_mesh directly.

I want to delete this endpoint. It's deprecated.

File: docs/multi_process.md (hunk 3 above)
Comment (olupton) [reply docs/multi_process.md]: I see people doing reshapes of `jax.devices()` in the wild, so it seems useful to say "don't do that". I can rework it to contain fewer details.

File: jax/_src/mesh_utils.py (hunk 1 above)
Comment (olupton) [reply jax/_src/mesh_utils.py:849]: OK, should we add a deprecation notice then? I don't see one.

Comment (yashk2810) [reply jax/_src/mesh_utils.py:849]: That can be a separate CL. For now, remove this?

**Review by yashk2810 — CHANGES_REQUESTED** [review]

File: jax/_src/mesh_utils.py (hunk 1 above)
Comment (yashk2810) [reply jax/_src/mesh_utils.py:849]: Or maybe this is fine.

**Review by yashk2810 — APPROVED** [review]

**Review by yashk2810 — APPROVED** [review]
